### ISSUE TITLE
Does not work on new data sets

### ISSUE THREAD

**Issue by huhuhu129** [Center-for-Health-Data-Science/multiDGD#20]
This is a very interesting job. However, when I applied it to more data sets, I found that it didn't work properly. I checked the reason for its error reporting. It seems that the loss was already very low in the first batch, while in the second batch, the loss would return a NaN value.@emdann @viktoriaschuster

**Comment by viktoriaschuster** [comment]
Hi @huhuhu129 

Thanks for telling us about this problem. Could you share a little information about the type and size of the data that you are trying to model? Are you using raw counts? I hope I can help and figure out how to prevent such issues in the future.

Best,
Viktoria

**Comment by huhuhu129** [comment]
Thank you for your reply. The data I am using is the 10X_PBMC data, which is downloaded from this linkhttps://scglue.readthedocs.io/zh-cn/latest/data.html . For the scRNA-seq data, we used 2,000 highly variable genes, and for the scATAC-seq data, we used 30,000 highly variable regions and analyzed it using LSI (following the Seurat v3 approach). Later, when I was debugging, I found that the following first piece of code function would return a large number of NaN values when obtaining the model loss, as shown in the second piece of output code below. This led to the appearance of extremely small values when calculating the loss, so NaN was returned all the time.
`def logNBdensity(k,m,r):
    ''' 
    Negative Binomial NB(k;m,r), where m is the mean and k is "number of failures"
    Here, k is the observed count and m is the predicted mean of the NB. r is the dispersion parameter
    k, and m are tensors of same shape
    r is tensor of shape (1, n_genes)
    Returns the log NB in same shape as k
    '''
    # remember that gamma(n+1)=n!
    eps = 1.e-10 # this is under-/over-flow protection
    x = torch.lgamma(k+r)
    x -= torch.lgamma(r)
    x -= torch.lgamma(k+1)
    inverse_rme = 1/(r+m+eps)
    x = x + k.mul(torch.log(m*inverse_rme+eps))
    x = x + r*torch.log(r*inverse_rme)
    print(x)
    return x`
`tensor([[nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        ...,
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan]], device='cuda:0',
       grad_fn=<AddBackward0>)
Training:  12%|█▏        | 60/500 [02:59<21:57,  2.99s/epochs, loss=nan, reconstruction_loss=nan]
finished training after  60  epochs according to early stopping based on  loss  for  10  epochs
[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]
#######################
Training status
#######################
True`

**Comment by huhuhu129** [comment]
> Thank you for your reply. The data I am using is the 10X_PBMC data, which is downloaded from this linkhttps://scglue.readthedocs.io/zh-cn/latest/data.html . For the scRNA-seq data, we used 2,000 highly variable genes, and for the scATAC-seq data, we used 30,000 highly variable regions and analyzed it using LSI (following the Seurat v3 approach). Later, when I was debugging, I found that the logNBdensity function would return a large number of NaN values when obtaining the model loss, as shown in the second piece of output code below. This led to the appearance of extremely small values when calculating the loss, so NaN was returned all the time. 
`tensor([[nan, nan, nan, ..., nan, nan, nan], [nan, nan, nan, ..., nan, nan, nan], [nan, nan, nan, ..., nan, nan, nan], ..., [nan, nan, nan, ..., nan, nan, nan], [nan, nan, nan, ..., nan, nan, nan], [nan, nan, nan, ..., nan, nan, nan]], device='cuda:0', grad_fn=<AddBackward0>) Training: 12%|█▏ | 60/500 [02:59<21:57, 2.99s/epochs, loss=nan, reconstruction_loss=nan] finished training after 60 epochs according to early stopping based on loss for 10 epochs [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan] ####################### Training status ####################### True`
Thans,
Hu

**Comment by viktoriaschuster** [comment]
I have never tried the model on such few variables. I will try to reproduce your error and get back to you as soon as possible.

**Comment by viktoriaschuster** [comment]
Although I do not have the exact same dataset because of your feature selection of the original files, I ran the model with feature selection to the same number of genes and peaks based on variance. I could not replicate your error. I uploaded a notebook with my data processing and training code [here](https://github.com/Center-for-Health-Data-Science/multiDGD/blob/feature_selection_tutorial/tutorials/example_feature_selection.ipynb). Maybe this will help you successfully train the model. If the problem still persists, please check the package requirements and let me know if it is still not fixed.

**Comment by huhuhu129** [comment]
Thank you for your reply and the experimental code. After abandoning our previous data-processing method and using the code you provided, the model is running normally. I'm still trying to figure out what the problems were with our previous data-processing approach, as well as the differences between the two data pre-processing methods. Many thanks for your help!

**Comment by viktoriaschuster** [comment]
I hope you will find out what the problem is. If there is anything that makes multiDGD finicky, please let us know :)
